checks.d/process.py

```diff
@@ -7,170 +11,213 @@
 from util import Platform
 
 
+DEFAULT_AD_CACHE_DURATION = 120
+DEFAULT_PID_CACHE_DURATION = 120
+
+
+ATTR_TO_METRIC = {
+    'thr':              'threads',
+    'cpu':              'cpu.pct',
+    'rss':              'mem.rss',
+    'vms':              'mem.vms',
+    'real':             'mem.real',
+    'open_fd':          'open_file_descriptors',
+    'r_count':          'ioread_count',  # FIXME: namespace me correctly (6.x), io.r_count
+    'w_count':          'iowrite_count',  # FIXME: namespace me correctly (6.x) io.r_bytes
+    'r_bytes':          'ioread_bytes',  # FIXME: namespace me correctly (6.x) io.w_count
+    'w_bytes':          'iowrite_bytes',  # FIXME: namespace me correctly (6.x) io.w_bytes
+    'ctx_swtch_vol':    'voluntary_ctx_switches',  # FIXME: namespace me correctly (6.x), ctx_swt.voluntary
+    'ctx_swtch_invol':  'involuntary_ctx_switches',  # FIXME: namespace me correctly (6.x), ctx_swt.involuntary
+}
+
+
 class ProcessCheck(AgentCheck):
+    def __init__(self, name, init_config, agentConfig, instances=None):
+        AgentCheck.__init__(self, name, init_config, agentConfig, instances)
+
+        # ad stands for access denied
+        # We cache the PIDs getting this error and don't iterate on them
+        # more often than `access_denied_cache_duration`
+        # This cache is for all PIDs so it's global
+        self.last_ad_cache_ts = 0
+        self.ad_cache = set()
+        self.access_denied_cache_duration = int(
+            init_config.get(
+                'access_denied_cache_duration',
+                DEFAULT_AD_CACHE_DURATION
+            )
+        )
 
-    SOURCE_TYPE_NAME = 'system'
-
-    PROCESS_GAUGE = (
-        'system.processes.threads',
-        'system.processes.cpu.pct',
-        'system.processes.mem.rss',
-        'system.processes.mem.vms',
-        'system.processes.mem.real',
-        'system.processes.open_file_descriptors',
-        'system.processes.ioread_count',
-        'system.processes.iowrite_count',
-        'system.processes.ioread_bytes',
-        'system.processes.iowrite_bytes',
-        'system.processes.voluntary_ctx_switches',
-        'system.processes.involuntary_ctx_switches',
-    )
-
-    def find_pids(self, search_string, exact_match, ignore_denied_access):
+        # By default cache the PID list for a while
+        # Sometimes it's not wanted b/c it can mess with no-data monitoring
+        # This cache is indexed per instance
+        self.last_pid_cache_ts = {}
+        self.pid_cache = {}
+        self.pid_cache_duration = int(
+            init_config.get(
```

Comment:
I think in most cases, this shouldn't be changed so I don't think it's worth the trouble for an instance parameter.